### DISCUSSION TITLE
SQLAlchemy using the same session across threads in the same async function

### DISCUSSION THREAD

**Discussion by ljiatu** [sqlalchemy/sqlalchemy#5838] [Usage Questions] [answered]
**Describe your question**
The question basically boils down to whether the following is safe, assuming there is a new SQLAlchemy session s being created every time this function is invoked:

```
async def func(s: Session):
  loop = asyncio.get_running_loop()
  await loop.run_in_executor(None, some_func, s)
  await loop.run_in_executor(None, some_other_func, s)
  ...
```
I'm thinking that it is safe because even though the session is used in (potentially) two different threads in the two run_in_executor calls, those two calls are not concurrent, since the second call won't be executed until the first one completes. This means the session is not accessed concurrently. Am I right in thinking that?

I asked the same question on Stackoverflow here: https://stackoverflow.com/questions/65650198/sqlalchemy-using-the-same-session-across-threads-in-the-same-async-function. But I feel like I may be able to get a faster response here.

Thanks so much!

**Comment by zzzeek** [comment, marked as answer]
hey there -

yes your thinking is correct, the Session will be used in threads but as long as it isn't used *concurrently* in more than one thread (or greenlet, or whatever), it should be fine.     

if you are using the SQLite backend, there are some extra things to take into consideration, you probably have to set the "check_same_thread" flag to False, there's some background on this at https://docs.sqlalchemy.org/en/13/dialects/sqlite.html#threading-pooling-behavior .

**Comment by zzzeek** [comment]
also, note that any objects that you've loaded or persisted with that Session are also part of that Session's state.    so if you had loaded some objects from that Session you want to make sure none of those are shared to concurrent threads either, they should only be used where that Session is currently being used.

**Comment by ljiatu** [comment]
Wow thanks @zzzeek for the lightning-fast response! I'm using Postgres as the backend so I guess I'm safe. Glad to know that my thinking is correct!

**Comment by CaselIT** [comment]
Note that this is valid only for the normal SQLAlchemy sessions, not the new AsyncSession introduced by the asyncio extension in 1.4. In that case it's in general not safe to share the same session (and engine) across different event loops.

I also think we can move do discussion once they re-add the feature

**Comment by ljiatu** [comment]
@CaselIT Thx for the reminder. Is it not safe to share the async session across different event loops, or different threads? In the example I had in the question, the (sync) session is accessed by the same event loop, but potentially different threads.

**Comment by CaselIT** [comment]
The sync connection does not care about event loops, so anyone is fine.

> Is it not safe to share the async session across different event loops, or different thread

I think it only on different loop, but I don't know if a loop can be used by multiple threads.
The limitation is actually of asyncio, mainly that some data structure, like the queue used to manage connections, binds to a specific loop, so trying to use an async connection to another loop can return an error stating that a task uses a different loop.
See https://github.com/sqlalchemy/sqlalchemy/issues/5823 for more details

**Comment by ljiatu** [comment]
Ok, thx for the explanation. I don't think we are going to use the same async session or connection on different event loops once we adopt 1.4. From what I understand the web framework we are using won't create two event loops.
